spec-docs/flexible-java-types.md

```diff
@@ -139,4 +139,214 @@ The compiler issues warnings specific to `@Nullable`/`@NotNull` in the following
  - a `@Nullable` value is assigned to a not-null location (including passing parameters and receivers to functions/properties);
  - a nullable value is assigned to a `@NotNull` location;
  - a `@NotNull` value is dereferenced with a safe call (`?.`), used in `!!` or on the left-hand side of an elvis operator `?:`;
- - a `@NotNull` value is compared with `null` through `==`, `!=`, `===` or `!==`
\ No newline at end of file
+ - a `@NotNull` value is compared with `null` through `==`, `!=`, `===` or `!==`
+
+## More precise type information from annotations
+
+Goals:
+ - Catch more errors related to nullability in case of Java interop
+ - Keep all class hierarchies consistent at all times (no hierarchy errors related to incorrect annotations)
+ - (!) If the code compiled with annotations, it should always compile without annotations (because external annotations may disappear)
+
+ This process never results in errors. On any mismatch, a bare platform signature is used (and a warning issued).
+
+### Annotations recognized by the compiler
+
+- `org.jetbrains.annotations.Nullable` - value may be null/accepts nulls
```

Comment:
I re-checked, and those differences (mostly in what NotNull means) seem to be irrelevant, so I added a list of other annotation packages below